### ISSUE TITLE
regression: incorrect jordan_form for matrix with repeated pure imag eigenvalues

### ISSUE THREAD

**Issue by cbm755** [sympy/sympy#15858]
Consider
```
>>> A = Matrix([[1, 1, 1, 0], [-2, -1, 0, -1], [0, 0, -1, -1], [0, 0, 2, 1]])

⎡1   1   1   0 ⎤
⎢              ⎥
⎢-2  -1  0   -1⎥
⎢              ⎥
⎢0   0   -1  -1⎥
⎢              ⎥
⎣0   0   2   1 ⎦


>>> A.eigenvals()

{-ⅈ: 2, ⅈ: 2}

```
Note two repeated eigenvalues

Now
```
>>> J = A.jordan_form(calc_transform=False)


⎡-ⅈ  1   0  0  0  0  0  0  0  0⎤
⎢                              ⎥
⎢0   -ⅈ  0  0  0  0  0  0  0  0⎥
⎢                              ⎥
⎢0   0   ⅈ  1  0  0  0  0  0  0⎥
⎢                              ⎥
⎢0   0   0  ⅈ  1  0  0  0  0  0⎥
⎢                              ⎥
⎢0   0   0  0  ⅈ  1  0  0  0  0⎥
⎢                              ⎥
⎢0   0   0  0  0  ⅈ  0  0  0  0⎥
⎢                              ⎥
⎢0   0   0  0  0  0  ⅈ  1  0  0⎥
⎢                              ⎥
⎢0   0   0  0  0  0  0  ⅈ  0  0⎥
⎢                              ⎥
⎢0   0   0  0  0  0  0  0  ⅈ  1⎥
⎢                              ⎥
⎣0   0   0  0  0  0  0  0  0  ⅈ⎦

```
This is the wrong size (and takes a long time).

If you ask for `J,P = A.jordan_form(calc_transform=True)`, it takes *very* long time and eventually errors out:
```
   1610                 # generalized eigenspace sharing the same eigenvalue.
   1611                 vec = pick_vec(null_small + eig_basis, null_big)
-> 1612                 new_vecs = [(eig_mat(eig, i))*vec for i in range(size)]
   1613                 eig_basis.extend(new_vecs)
   1614                 jordan_basis.extend(reversed(new_vecs))

TypeError: unsupported operand type(s) for *: 'MutableDenseMatrix' and 'NoneType'
```

Seems quite similar to #9274 which I supposedly fixed, *sigh*.  It seems to work on Diofant, so we should look at how they fixed #9274...

**Comment by cbm755** [comment]
Its a regression, 97aeed33d99fabd4975c8e4fa6b322b4f87072b5 @siefkenj

**Comment by siefkenj** [comment]
This appears to be due to `.rank` being computed incorrectly.

```
In [20]: A = Matrix([[1, 1, 1, 0], [-2, -1, 0, -1], [0, 0, -1, -1], [0, 0, 2, 1]
    ...: ])

In [21]: B = A - Matrix.eye(4)*I

In [22]: B.rank()
Out[22]: 3

In [23]: (B**2).rank()
Out[23]: 2

In [24]: (B**3).rank()
Out[24]: 3

In [25]: (B**3).rank(simplify=True)
Out[25]: 3

In [26]: simplify(B**3).rank()
Out[26]: 2
```

**Comment by sidhantnagpal** [comment]
Resolved in #15875 and #15882.
